### PR TITLE
chore(types): remove optional types on ConfigFlags interface

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
Removing the optional types on `ConfigFlags` lets us remove a few null
suppression errors here and there and removes some `strictNullChecks`
violations as well.

To make it easy to create a correctly-formed `ConfigFlags` object this
also introduces a `createConfigFlags` helper function which takes a
partial `ConfigFlags` object and returns one with default properties
set.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Unit tests (`npm test`) were run locally and passed
- [x] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

GitHub Issue Number: N/A

## What is the new behavior?

Basically this just cleans up the typing on the `ConfigFlags` object a little bit and introduced a helper function to make it easy to create correctly typed objects.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

I just made sure everything is compiling and tests are passing.

### CODE REVIEW

File: src/cli/config-flags.ts [hunk 1]
```diff
@@ -269,8 +269,27 @@ export interface ConfigFlags
...
 }
+
+/**
+ * Helper function for initializing a `ConfigFlags` object. Provide any overrides
+ * for default values and off you go!
+ *
+ * @param init an object with any overrides for default values
+ * @returns a complete CLI flag object
+ */
+export const createConfigFlags = (init: Partial<ConfigFlags> = {}): ConfigFlags => {
```
Comment (rwaskiewicz) [inline src/cli/config-flags.ts:285]: I wonder if it makes sense to have a separate one of these for creating the config flags for tests.  IDK if reasonable defaults for tests will always be the reasonable defaults we use in `src` 🤔 Thoughts?

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
first pass looks good - I still need to go through the tests and verify calling the helper function with defaults doesn't drop any coverage (where adding new fields could theoretically change the codepath taken).  A few questions/comments in the mean time

File: src/cli/test/run.spec.ts [hunk 2]
```diff
@@ -129,6 +138,9 @@ describe('run', () => {
       validatedConfig = mockConfig(sys);
       validatedConfig.outputTargets = [];
 
+      console.log('run.spec::beforeEach');
+      console.log(validatedConfig.flags);
+
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/cli/test/run.spec.ts:143]: sigh...I should read things more closely!

File: src/cli/config-flags.ts (hunk 1 above)
Comment (alicewriteswrongs) [reply src/cli/config-flags.ts:285]: we could have a separate one for tests if we wanted to, my intention here was to create an object which is the minimum which complies with the interface so I think we'd want _that_ in both test code and, uh, actual code, but if we wanted to have additional defaults for tests we could have a separate function for that, although I think at present it would look something like this:

```ts
const mockConfigFlags = (overrides: Partial<ConfigFlags> = {}): ConfigFlags => {
  const init = createConfigFlags();

  return { ...init, ...overrides };
}
```

File: src/compiler/compiler.ts [hunk 3]
```diff
@@ -18,7 +18,7 @@ import ts from 'typescript';
  * @returns a new instance of a Stencil compiler
  * @public
  */
-export const createCompiler = async (userConfig: Config): Promise<Compiler> => {
+export const createCompiler = async (userConfig: ValidatedConfig): Promise<Compiler> => {
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/compiler.ts:21]: that makes sense - I think I was playing around with this and meant to leave a 'do we want to do this?' comment

File: src/cli/config-flags.ts (hunk 1 above)
Comment (rwaskiewicz) [reply src/cli/config-flags.ts:285]: Yeah - TBH I waffle on what the best thing to do is here. On one hand, it's nice to have a clear distinction between the two - one for testing, one for the impl. On the other, we'd probably end up writing exactly what you have there, which still uses the underlying version for tests.  I say let's let YAGNI ride and keep things as is for now

**Review by rwaskiewicz — APPROVED** [review]
